### DISCUSSION TITLE
onSuccess doesn't get called when query is retrieved from cache

### DISCUSSION THREAD

**Discussion by georgiosd** [TanStack/query#937] [General]
Hi!

I used apollo previously on a client project as per the client's request and I really regretted ever doing that. The experience was dreadful, especially when dealing with the cache.

react-query which I then used by my own free will on another project :D has been a breath of fresh air.

However, I do find strange that the `onSuccess` callback is not getting called when the result is retrieved from cache. It's a bit limiting when you try to use `onSuccess` and `onError` to set state.

Could we add that in somehow?

PS: I understand that it's expected behavior based on what the documentation says - but it could be clearer in the docs.

**Comment by ryantomaselli** [comment]
I would also like to set state onSuccess even when the data is cached.

I'm setting state such that I can differentiate between resting, loading, fetching more and filtering...if I don't get a callback when the query hits the cache I can't set the state back to resting.

I should add that I too am enjoying hooking things up with react-query :)

**Reply by TkDodo** [reply]
> It's a bit limiting when you try to use onSuccess and onError to set state.

> I would also like to set state onSuccess even when the data is cached.

I’m not sure which state you are setting? react-query generally manages state (data, error) and “loading-states” (isLoading, isError) for you. Not sure what the “resting” state would be in your example.

I for example use the `onError` callback to show an Error toast in the ui, and I would be extremely unhappy if this would happen on every render cycle, which is effectively what would happen if cache hits would also call trigger callbacks...

**Reply by georgiosd** [reply]
I see your point.

In the way that we build apps, sometime we have to set some other state of our own using the `onSuccess` callback. This is not possible right now if the query is cached, we must then also check the cache and do it with `useEffect`

You don't necessarily have to call `onSuccess` on every cache hit, but at least on every key change. That would obviously be a breaking change, but it could be a drop-in setting.

What do you think?

**Reply by TkDodo** [reply]
> In the way that we build apps, sometime we have to set some other state of our own using the onSuccess callback.

Do you have an example for that, because it sounds a bit like state duplication. I would just do useQuery again in the components that need this “other state”, and compute the state they need on the fly (if expensive, either with useMemo or with the new `select` queryOption in v3).

> You don't necessarily have to call onSuccess on every cache hit, but at least on every key change.

I think this should be the case already. When the key changes, it means a new entry in the queryCache is created, which means data is re-fetched per default (you can tweak that of course). Again, a concrete example would help in understanding the problem.

**Reply by georgiosd** [reply]
Network requests have related state, not all state is the network result it self.

One recent example is a widget where you input a URL and fetches some information from that page. On success, I am updating some validation and invoking a callback on the parent with it. 

With the current design, I have to provide both an `onSuccess` callback but also watch for a cache hit different to the last known value and invoke the callback.

**Reply by TkDodo** [reply]
> Network requests have related state, not all state is the network result it self.

yes, derived state can be implemented with a custom hook or with the `select` option:

```
const useMyDerivedQuery = () => {
    const { data} = useQuery(...)

   return data.filter(...).map(...)
}
```

> One recent example is a widget where you input a URL and fetches some information from that page. On success, I am updating some validation and invoking a callback on the parent with it.

Aren't these things dependent on the data rather than dependent on where the data comes from (network  vs cache)? An effect that executes whenever the data changes, for example:

```
const { data } = useQuery(...)

React.useEffect(() => {
    updateSomeValidation(data)
    invokeSomeParentCallback(data)
}, [data, invokeSomeParentCallback, updateSomeValidation])
```

**Reply by georgiosd** [reply]
The first may be an overkill for some things because it may not just be a transformation but say setting three other states depending on content.

The second is much closer to what I ended up doing, with the caveat that it's not supported for calls which return arrays.

I did however think that I could do the `useEffect` on the key + `isSuccess` - perhaps the most minimal change to make this scenario easier would be to re-return the `key` from `useQuery`?

**Reply by TkDodo** [reply]
> setting three other states depending on content.

yes, I would do that in an effect. But I have rarely needed it.

> with the caveat that it's not supported for calls which return arrays.

why not? React-query has structural sharing, so you will only get a new array if really anything changes in the data, so your effects will only run if data has changed.

> I did however think that I could do the useEffect on the key + isSuccess - perhaps the most minimal change to make this scenario easier would be to re-return the key from useQuery?

yes, key + isSuccess should also work, but I think depending on the data would be better.

**Reply by georgiosd** [reply]
That's a good point. We got stuck thinking that arrays shouldn't be used as dependencies because the comparison is shallow but the array reference gets replaced anyway if you create a new array with more/less items.

Hm.... ok. I guess my question now is - what are good use cases for `onSuccess`?

**Reply by TkDodo** [reply]
> Hm.... ok. I guess my question now is - what are good use cases for onSuccess?

good question, I have never needed it on `useQuery` 😅 . I am using `onError` to display error toasts, which is an imperative action:

```
onError: (error) => showToast(error.message)
```

and sometimes `onSuccess` / `onSettled` on `mutations`, for example, to close a modal:

```
<button onClick={() => mutate(variables, { onSettled: () => modal.close() } } />
```

**Reply by georgiosd** [reply]
💯 Thanks!

**Reply by christopher-caldwell** [reply]
@TkDodo Your explanation was very helpful. Thanks! I had the same question, and was able to work it out based on your answers.

**Reply by aakashsrdev** [reply]
Thanks @TkDodo

**Reply by varian97** [reply]
@TkDodo Sorry to continue this thread 4 years later, but I meet the same problem today.
In my case, I have an API to retrieve shipping fee from the backend, so I use useQuery for that. When the query succeed, I want to store it to redux. So I use `onSuccess` for that

```javascript
const dispatch = useDispatch()
const { isLoading, isFetching } = useQuery(['shipping-fee'], loadShippingFeeAPI, {
  onSuccess: (res) => dispatch(setShippingFeeData(res))
})
```

However the issue arise if the customer go to the other page (ex: they press back button) when the API call still not finished, let me breakdown it into steps:
1. Customer open the checkout page, the useQuery for loading shipping fee called, the API call to backend called
2. Before the api call finished, customer press back button, going to the previous page (let say the product detail page)
3. Customer now on the product detail page, the API call finished with status code 200
4. Now at this state, the redux still empty, I assume because onSuccess not called. But the shipping fee query already cached by react-query
5. When customer go to the checkout page again, the UI not updated with the shipping fee info because it reads the data from redux. useQuery also does not call the API again because it is already in the cache

I do solve this problem by using useEffect on the client component, but the questions are:
1. Does this happened because onSuccess / onError does not called when user already changing page? why the data cached inside react-query but the code that dispatch data to redux not called?
2. What are the real use case of `onSuccess` / `onError` then? Everything can be put inside useEffect, including showing error toast / success toast

Thank you

**Reply by TkDodo** [reply]
> When the query succeed, I want to store it to redux

That's your problem. You don't want that. Don't do this.

> When the query succeed, I want to store it to redux

There are none. We have removed those callbacks in v5.

**Reply by varian97** [reply]
Thanks @TkDodo 
I read your article https://tkdodo.eu/blog/breaking-react-querys-api-on-purpose and it shows exactly the same problem that I wrote.

**Comment by cliffordfajardo** [comment]
Adding my story as a data point in case others stumble upon this behavior as well.
My code was relying on the `onSuccess` callback of a `useQuery` hook to call a `setState`. It looked something like this:

```tsx
const SomeComponent = ({ children }: { children: ReactNode }) => {
  const [renderItems, setRenderItems] = useState([])
   useQuery(..., {
    onSuccess(APIData){
       // 🔴  onSuccess will not trigger if data is grabbed from cache
       const renderData = util.transform(APIData); // all this does is transform the API data into another format
        setRenderItems(renderData) // so of course this doesn't trigger and my component below won't re-render with new daat
    }
  })
  
  return (
     <SomeListComponent items={renderItems} />
  )
}
```

When I enabled caching (_this applies to the in memory `cacheTime` and `persistQueryClient` approaches_)  I was confused as to why my `onSuccess` callback wasn't getting called.  This was problematic because I needed my UI to re-render.

## Solution: 
- remove your `onSuccess` usage and derive your data instead. 

The key thing here is that since I am no longer using OnSuccess, I am able to get the data from the cache.

```tsx
const SomeComponent = ({ children }: { children: ReactNode }) => {
  const {data: APIData} = useQuery(...)  
  const renderItems = APIData ? util.transform(APIData): []; ✅  works
  return (
     <SomeListComponent items={renderItems} />
  )
}
```

**Reply by TkDodo** [reply]
we will make it more clear in the docs that onSuccess is tied to data fetching, like onError. To perform side effects whenever data changes, a useEffect is still the best approach.
I also think it’s sub-optimal to do data transformations and then store the result in state, as this violates a principle of react to only store the most minimal state and derive the rest. The `select` option offered by RQ is perfect for that scenario.

**Reply by cliffordfajardo** [reply]
Thanks -  I’ll start looking at select for the data transformation 👀

**Reply by davit-b** [reply]
Yep, ran into this same problem and took an hour to realize that onSuccess is NOT called on cache hits.

**Comment by donysukardi** [comment]
I'm hoping to have some callback regardless of whether it's freshly fetched or retrieved from cache as well. My use is case is to trigger some side effect, e.g. redirect with some default params

**Reply by cliffordfajardo** [reply]
> we will make it more clear in the docs that `onSuccess` is tied to data fetching, like onError. To perform side effects whenever data changes, a useEffect is still the best approach. I also think it’s sub-optimal to do data transformations and then store the result in state, as this violates a principle of react to only store the most minimal state and derive the rest. The `select` option offered by RQ is perfect for that scenario. - @TkDodo 

I've had this thought since I started using react-query 2 years ago; Probably good 3-4 times where I had this scenario show up, I spend 45 minutes tracing my code only to remember that `onSuccess` is tied to data fetchin & doesnt get called on Cached Data
I always thought it was strange that `onSuccess` doesnt get called if the data is grabbed from the cache;  

As @donysukardi said, sometimes you do just want to fire a side effect in the `onSuccess` regardless of whether or not the data was in the cache or note

@TkDodo - when I think of `onSuccess` from a library consumer point of view, I think:
1. ✅  onSuccess = my data was able to be fetched successfully from some source (no-cache) -- this is currently how RQ works
2. 🟡 onSuccess = my data was able to be fetched successfully from some source (could be my cache)

✅  -how it works
🟡 - how i wish it worked


## Proposal
<table>
<tr>
<td> <b>current `onSuccess` behavior</b> </td> <td> <b>What I wish happened</b> </td>
</tr>
<tr>
<td>

```tsx
const SomeComponent = () => {
  const [renderItems, setRenderItems] = useState([])
   useQuery(..., {
    onSuccess(APIData){
       // ✅ the first time this will work
        setRenderItems(APIData)
    }
  })
  return (
    <SomeListComponent items={renderItems} />
  )
}

const SomeComponent = () => {
  const [renderItems, setRenderItems] = useState([])
   useQuery(..., {
    onSuccess(APIData){
       // ❌ doesnt work if data is in cache
        setRenderItems(APIData)
    }
  })
  return (
    <SomeListComponent items={renderItems} />
  )
}
```

</td>
<td>

```tsx
const SomeComponent = () => {
  const [renderItems, setRenderItems] = useState([])
   useQuery(..., {
    onSuccess(APIData){
       // ✅ should work the same first time & if we have data in cache
        setRenderItems(APIData)
    }
  })
  return (
    <SomeListComponent items={renderItems} />
  )
}
```

</td>
</tr>
</table>

As a user, I don't care too much that this data came back from the cache or not, to me if I RQ has the data and is able to retrieve from some source that is a "success". When I first used `onSuccess` thats how I assumed it should work.

I think @donysukardi is thinking the same - please let us know Dony

**Reply by TkDodo** [reply]
for v5, I'm thinking of adding an `onDataChanged` callback that will be called with current and previous data:

```
onDataChanged: (apiData) => ...
```

I think this would work for you use-case, I'd just need to figure out the timing of when exactly to call this. For example, if the observer key changes, we instantly give the data back from the cache if it exists. But we can't call the callback, because this can happen _during_ rendering. So essentially, we'd have to spawn a `useEffect` ourselves 😅 , but then this callback wouldn't live in the query/core. I would prefer it if we'd still leave this to user-land.

**Reply by dima-getselectstar** [reply]
I find this feature very useful. This is so weird to use `useEffect` whenever we need to update some state, based on data we received (from cache or network).  I also prefer if `onSuccess` gets called each time data were retrieved doesn't matter from where. However, `onDataChanged` also sounds like a good enough alternative.

**Reply by georgiosd** [reply]
I'm very curious about this @dima-getselectstar - why is it weird to use `useEffect` to monitor data changes?

**Reply by ShamansCoding** [reply]
@georgiosd  It is not good because the developer as a user of a given API should bother whether or not data is retrieved from the cache or the server. It is a leak in useQuery abstraction in my opinion and in some cases, I can rely on onSuccess callback if data is retrieved from the server every time, and in some cases, it can not because it's taken from the cache.

Moreover, I can not create my abstractions on top of useQuery to hide some details from external users. If I will create the hook usePosts which will return a collection of posts user of my hook can not rely on onSuccess if he doesn't know if the cache is enabled inside of this hook. If the cache is disabled he can put some logic inside the onSuccess callback and be sure, that every time the hook is returning some data onSuccess will fire. But if the cache is enabled he can't do so. So he needs to go inside usePosts realization and see things, that he doesn't need to know most of the time as usePosts  API user.

**Reply by georgiosd** [reply]
That seems to be a confusion around what `onSuccess` does, rather than anything else. The documentation is clear about what it does and when it should be used. We can argue that's a bad name for what it does but it can be argued both ways. 

If you expect that you need to create an abstraction over `useQuery` which invokes a callback every time the data changes, what prevents you from doing that using `useEffect`?

```
function usePosts(onDataChanged) {
  const query = useQuery(...)

  useEffect(() => {
    onDataChanged(query.data)
  }, [query.data])

  return query
}
```

Sure, maybe adding the `onDataChanged` callback above would alleviate that from you, but so would a no-code solution 😄 I'm teasing, but you get my drift. Abstractions are designed to support many use cases with their own considerations, not just our own. 

In this particular case, I think it's a very good design in that the expensive operation is the network, so it's rightfully given a special handler. I don't see a leak anywhere, only programmers who get lazy reading the docs 😆 I was one of them, trust me, I started the thread 🤣 

PS: I am not sure it's a good idea for a developer to design their calls to `react-query` or any other library for that matter, assuming a particular settings is on or off (in this case, the cache).

**Reply by ShamansCoding** [reply]
In general, nothing prevents me from doing so. We can often come up with some kind of "hack" to solve our problems if something does not behave as expected. I totally agree with @[cliffordfajardo](https://github.com/cliffordfajardo) that the onSuccess behavior is frustrating for the user of the hook API and behaves completely differently than you would expect. Even if you have read the documentation, this moment slips away with time and there is an opportunity to run into it again at the most unexpected moment.

After all, when we get "data" directly from the useQuery hook, the abstraction allows us not to think about where the data came from, so why in the case of onSuccess do we have to think about it? This is weird. The data has arrived, so the data has arrived. And from where useQuery decided to take it under the hood should remain as hidden realization detail. Just like it is with returned "data".

**Reply by cliffordfajardo** [reply]
I've come back to this thread a few times, because of I assumed `onSuccess` would get called if react-query has the data
its not intuitive that `onSuccess` doesn't get called when the data is in the cache. `onSuccess` is `onSuccess` lol! 😄 no matter if the data came from a cache or not

> This is so weird to use useEffect whenever we need to update some state, based on data we received (from cache or network). I also prefer if onSuccess gets called each time data were retrieved doesn't matter from where

+1 to previous thread reply

**Reply by TkDodo** [reply]
> I also prefer if onSuccess gets called each time data were retrieved doesn't matter from where

I thought about this `onDataChanged` callback some more and I don't think we'll do it. Let me explain:

When data is read from the cache, you get it back _synchronously_. There is no way for us fire a "side-effect" here, apart from manually triggering a `useEffect`. However, `useEffect` is react specific, so we would need to re-create this behaviour for every framework adapter. And even if we do this, the root problem with effects still persist that there will be _one_ render cycle in between where data will be out-of-sync. So you will have `data` returned from `useQuery` already be the "new" data read from the cache, but your local "copy" of it will still hold the "old" data.

`onSuccess` works now because data fetching already doesn't happen during rendering, so whenever data comes back, we can trigger side-effects.

FWIW, I think this is an anti-pattern that should be avoided:

```
const [renderItems, setRenderItems] = useState([])
useQuery(..., {
  onSuccess(APIData){
    setRenderItems(APIData)
  }
})
```

it _duplicates_ server state into client state which shouldn't be necessary. This sentence says it best:

> this is so weird to use useEffect whenever we need to update some state, based on data we received (from cache or network)

Imo, you _should_ be feeling weird, because it is a weird thing to do. State should have a single source of truth, and further state can be _derived_. Having effects that sync state is, again, not a good idea. If we provide an abstraction to make you feel less weird, we would further encourage doing this, even though I feel strongly that you probably shouldn't. At least with a `useEffect` in user-land, you'll see explicitly what you are doing.

I would like to encourage everyone in this thread to read the excellent [You Might Not Need an Effect](https://beta.reactjs.org/learn/you-might-not-need-an-effect) section in the new react docs. If you really _must_ do this, the docs even say that setting state _during rendering_ would be preferred. This is the equivalent of the class based `getDerivedStateFromProps`:

```
const [renderItems, setRenderItems] = useState([])
const { data } = useQuery(...)

if (data !== renderItems) {
  setRenderItems(data)
}
```

The docs are still pretty clear that you probably don't want this though:

> Although this pattern is more efficient than an Effect, most components shouldn’t need it either. No matter how you do it, adjusting state based on props or other state makes your data flow more difficult to understand and debug.

I know that the example is contrived and you're all probably not setting data 1:1 into local state, but still:

- if the manipulation is _synchronous_, state should be derived during render (more on this here: https://tkdodo.eu/blog/dont-over-use-state)
- if the manipulation is _asynchronous_ - it should be _another_ query, from which you can then again synchronously derive things.
- if you need data to be put into state as an "initial value" for further updates by the user, you also don't want an effect, because that would potentially overwrite user-input. There are multiple ways to go about that, which I'm describing here: https://tkdodo.eu/blog/react-query-and-forms

**Reply by ShamansCoding** [reply]
> FWIW, I think this is an anti-pattern that should be avoided:
> 
> ```
> const [renderItems, setRenderItems] = useState([])
> useQuery(..., {
>   onSuccess(APIData){
>     setRenderItems(APIData)
>   }
> })
> ```

And if I need to display the data filtered by some condition along with the original data? For example, to show a general list in one part of the interface, and the same list in the other, but filtered by belonging to the user who is viewing the page.

```
const [itemsOfCurrentUser, setItemsOfCurrentUser] = useState([])
useQuery(..., {
  onSuccess(APIData){
      setItemsOfCurrentUser(APIData.filter(..some filter by user condition, may be by current userId...)) 
  }
})
```

I understand that we can do the same thing in other ways, but in this particular situation, it does not look like bad practice to do so.

And there are many such situations. For example, I can fire some kind of event to my clickstream database based on what is currently displayed on the user's page, from cache or not. Doing this in onSuccess looks like a very convenient solution. 

Don't get me wrong, react-query is by far the best library I've come across for working with back-end data. But things like working with onSuccess do cause some frustration.

In general, the scope of onSuccess is extremely narrow at the moment and I see it more like onServerDataLoadSuccess.

**Reply by TkDodo** [reply]
This isn't even about react-query. Your above example is really just:

```
const { data } = useQuery(...)
const itemsOfCurrentUser = data.filter(...)
```

this is a prime example of what I meant by `derived state`. It's all over the docs page I linked and the blog-posts I posted. You can wrap this in `useMemo` if you don't want to execute it during every render, but you should execute it _during render_, because that's where data should be derived. **Derived state isn't state**. With this, the `itemsOfCurrentUser` will always be in-sync with the server state returned from `useQuery`.

> In general, the scope of onSuccess is extremely narrow at the moment

I agree. I'm pondering to remove it 😅

**Reply by georgiosd** [reply]
+1 on removing it completely 😆

**Comment by miuan** [comment]
I see the problem when using the `onSuccess` to update redux-store

for ex: we have switch A x B,  what call some endpoint with A or B, and response we consume like `const { data } = useQuery('some-ep', {A|B})` bud for some components we also push to the store... so in case if you switch from A to B, all good , but if you switch back to A, then we have correct data, but store wasn't updated

only solution I have in mind I have to use useEffect to check if data was updated, but that is not good approach , duple every useQuery with useEffect

**Reply by TkDodo** [reply]
yes, the preferred solution is to not sync with redux at all :)

**Reply by miuan** [reply]
:) thank you for very quick answer, easy to say, harder to do :)

let say you have two components:
1. have Switcher A / B with show result - here is easy you can call useQuery directly base on user input and show the result
2. doesn't have Switch but only show the result , so you have two option somehow propagate user setting to call useQuery also here, or propagate data from useQuery called in first component to second, booth solution are harder by how far in DOM these two components are each other (ex: viewer is in header , switcher is deeply in view)

I very like concept of useQuery, but I just think we should consider another solution for retrieving data from server than useQuery in our app :(

**Reply by TkDodo** [reply]
This is a client state problem, not a server state one. Instead of putting the result from useQuery into redux, you want to put the user input itself into redux. This will give you something like:

```
const useResult = () => {
  const userInput = useSelector(state => state.userInput)
  return useQuery({
    queryKey: ['result', userInput],
    queryFn: () => fetchResult(userInput),
    enabled: !!userInput
  })
}
```

Now you can call `useResult()` wherever you want, your client state lives in redux, your server state lives in react-query, and you have no state duplication problem. Every time you update `userInput` in redux, your server state will automatically be up-to-date with that input.

**Reply by miuan** [reply]
I have also this in mind, but that is in my opinion more complicate regarding how far our app is already developed, but thank you very much

**Comment by ppYoung** [comment]
a common situation is: we use a store to share data among components, the store data comes from RQ and can be changed by user action

```javascript
const [complexData setComplexData] = useAtom(complexDataAtom)

const { data } = useQuery(…, {
    onSuccess(apiData) {
        setComplexData(apiData)
    }
})

// component A
setComplexData((draft) => draft.a = someUserValue)

// component B
setComplexData((draft) => draft.b = someUserValue)

``` 
but if switch the query key a -> b -> a, the store data cannot be set to correct data with queryKey a, still keep the previous data from queryKey b, the only way to keep the data update is `useEffect`, and unlike your demo about useResult when complexData changed to refetch data, no, i don’t need to refetch data when store change, the only solution is:
```javascript
useEffect(() => {
    setComplexData(data)
}, [data])
``` 

it’s so weird, i just need a callback when data fetched, no matter from server or from cache🥲

**Reply by TkDodo** [reply]
`useEffect` solves that. However, I truly believe that you shouldn't sync state from react-query to another state manager. That's the root cause of the problem, and it will yield other problems, e.g. user input being overwritten from background refetches.

**Reply by georgiosd** [reply]
It's amazing to me that 2.5 years later we're still discussing the same thing over and over 😆 

It's also amazing to me that you @TkDodo (still) have the patience to respond to people and haven't closed the thread!

**Reply by aakashsrdev** [reply]
Thanks @TkDodo @georgiosd @ShamansCoding for discussing this topic in so much detail. This use case of onSuccess was worth discussing. I've not got this particular use case yet, but I know now how to handle it now. 😃

**Comment by TkDodo** [comment]
> It's also amazing to me that you @TkDodo (still) have the patience to respond to people and haven't closed the thread!

closing discussions is a new thing, but I'll make use of it now. `onSuccess` will be removed in v5 so we won't have that problem:

- https://github.com/TanStack/query/discussions/5279

**Reply by aakashsrdev** [reply]
@TkDodo then what will replace this onSuccess callback ?

**Reply by TkDodo** [reply]
have you read the linked blogpost?

**Reply by aakashsrdev** [reply]
Sorry, got the point. Thanks for detailed blog

**Reply by davit-b** [reply]
Glad you are removing onSuccess - it's a courageous DX decision. So many people, me included, mis-used onSuccess for side effects.

And all the damn tutorials online teach you to use onSuccess for side-effects too :( 

Thanks for having the courage to make this change.

**Comment by ppYoung** [comment]
亲，你的邮件东阳已经收到了

**Comment by t-hoffman** [comment]


**Reply by TkDodo** [reply]
you shouldn’t perform side-effects inside `useMemo`, that breaks the rules of react. If you want cached sorting - just do it inside the queryFn, before returning the values:

https://tkdodo.eu/blog/react-query-data-transformations#1-in-the-queryfn

**Reply by t-hoffman** [reply]
Oh wow duh because that winds up going in the cache and only gets calculated again once the data is fetched again...... 😒 sorry it's late and I am tired, thank you so much for responding!

**Comment by lili21** [comment]
Here is my use case. I'm trying to collect some metrics , and measure the performance improvement from RQ.  

Beside, I'm using the persister with RQ. and It would be helpful If I can collect the cache rate metrics and adjust the cache strategy.
